examples/examples_test.go

```diff
@@ -0,0 +1,120 @@
+/*
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package examples
+
+import (
+	"io"
+	"os"
+	"path"
+	"path/filepath"
+	"testing"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/util/yaml"
+	"k8s.io/client-go/kubernetes/scheme"
+	"sigs.k8s.io/service-apis/api/v1alpha1"
+)
+
+func findExamples(t *testing.T, dirPath string) []string {
+	filenames := []string{}
+
+	t.Helper()
+
+	for _, pattern := range []string{
+		path.Join(dirPath, "*.yaml"),
+		path.Join(dirPath, "*.yml"),
+	} {
+		matches, err := filepath.Glob(pattern)
+		if err != nil {
+			t.Fatalf("glob %q failed: %s", pattern, err)
+		}
+
+		filenames = append(filenames, matches...)
+	}
+
+	return filenames
+}
+
+func decodeUnstructured(t *testing.T, filename string) []unstructured.Unstructured {
+	t.Helper()
+
+	f, err := os.Open(filename)
+	if err != nil {
+		t.Fatalf("%s", err)
+	}
+
+	defer f.Close()
+
+	objects := []unstructured.Unstructured{}
+	decoder := yaml.NewYAMLOrJSONDecoder(f, 4096)
```

Comment:
I think there's a better way to do this. It's been awhile since I've done this, but I think it can be something kinda sorta like this:

```golang
bytes, err := ioutil.ReadFile(filename)
if err != nil {
  t.Fatalf("Error reading file %s", err.Error())
}

scheme := runtime.NewScheme()
_ = v1alpha1.AddToScheme(scheme)
serializer.NewCodecFactory(v1alpha1.Scheme).UniversalDeserializer().Decode
obj, groupVersionKind, err := decode(bytes, nil, nil)

if err != nil {
  t.Fatalf("...")
}
```

That assumes that all our example files will only ever include Service APIs types and only include one type per YAML file. Of course we could add support for those cases in the future.

Reply:
> I think there's a better way to do this.

Does the decoder have different semantics than unmarshalling an unstructured?

> That assumes that all our example files will only ever include Service APIs types and only include one type per YAML file.

We currently have examples with more than one YAML object per file, so we would need to handle that. I've been assuming that we will end up with examples that also have core types (Services, Secrets, etc).

Reply:
We should try to reuse the k8s parser, it may lead to even more validation (e.g. the CRD validation). Isn't it possible to split the multifile YAML by the separator `---` without a YAML before running the above snippet?

Reply:
Does the UniversalDeserializer do more checking? I can easily turn the unstructured back into bytes, and deserialize it with an API that does more checking. I would also like to to the API server validations if I can find the right API packages :)

Reply:
I don't think the UniversalDeserializer will do that for us unfortunately. If we're strictly looking for better e2e tests, I think we could have a simple test that installed our CRDs in a cluster and then attempted to apply all the examples. That would have the benefit of actually running validations, but unfortunately would be difficult and/or not desirable to run locally. As far as parsing yaml files with multiple resources, there's this classic issue that has a bunch of [potential approaches for parsing YAML into Kubernetes resources](https://github.com/kubernetes/client-go/issues/193), including some that support multiple resources per file.

Reply:
@robscott Updated to use UniversalDeserializer :)